File: patterns/decorator.js
//decorator pattern
// add new features to existing objects

// Define the Component interface, which specifies the methods that ConcreteComponents and Decorators must implement
class Component {
    operation() { }
}

// Define the ConcreteComponent, which implements the Component interface and provides the core functionality
class ConcreteComponent extends Component {
    operation() {
        console.log("ConcreteComponent: Operation");
    }
}

// Define the Decorator, which implements the Component interface and adds behavior to the ConcreteComponent
class Decorator extends Component {
    constructor(component) {
        super();
        this.component = component;
    }

    operation() {
        this.component.operation();
    }
}

// Define ConcreteDecorators, which add specific behavior to the Component
class ConcreteDecoratorA extends Decorator {
    constructor(component) {
        super(component);
    }

    operation() {
        super.operation();
        console.log("ConcreteDecoratorA: Operation");
    }
}

class ConcreteDecoratorB extends Decorator {
    constructor(component) {
        super(component);
    }

    operation() {
        super.operation();
        this.addBehavior();
    }

    addBehavior() {
        console.log("ConcreteDecoratorB: Added behavior");
    }
}

// Use the Decorators to add behavior to the ConcreteComponent
const component = new ConcreteComponent();
const decoratorA = new ConcreteDecoratorA(component);
const decoratorB = new ConcreteDecoratorB(decoratorA);

decoratorB.operation();

/* In this example, we have a Component interface that specifies the methods that ConcreteComponents and Decorators must implement.We define a ConcreteComponent class that implements the Component interface and provides the core functionality.

We define a Decorator class that implements the Component interface and adds behavior to the ConcreteComponent.The Decorator class contains a reference to a Component object, which it decorates by calling the component's operation method and then adding its own behavior.

We define ConcreteDecorator classes that add specific behavior to the Component.Each ConcreteDecorator takes a Component object as a parameter and adds its own behavior to the decorated component's operation method.

We create instances of the ConcreteComponent and use ConcreteDecorator classes to add behavior to it.We can chain multiple decorators together to add multiple behaviors to the ConcreteComponent.

In this way, the Decorator pattern allows us to add behavior to objects dynamically at runtime by wrapping them in a series of Decorator objects.This can be useful in situations where we want to add behavior to an object without modifying its source code. */

File: patterns/decorator.py
# Define the Component interface, which defines the interface for the ConcreteComponent and Decorator classes
class Component:
    def operation(self):
        pass

# Define the ConcreteComponent class, which implements the Component interface and provides the basic behavior
class ConcreteComponent(Component):
    def operation(self):
        return "ConcreteComponent: Doing basic operation"

# Define the Decorator class, which implements the Component interface and adds new behavior to the ConcreteComponent class
class Decorator(Component):
    def __init__(self, component):
        self._component = component

    def operation(self):
        return self._component.operation()

# Define the ConcreteDecorator class, which extends the Decorator class and adds new behavior to the ConcreteComponent class
class ConcreteDecoratorA(Decorator):
    def operation(self):
        return f"ConcreteDecoratorA: Adding new behavior to {self._component.operation()}"

class ConcreteDecoratorB(Decorator):
    def operation(self):
        return f"ConcreteDecoratorB: Adding new behavior to {self._component.operation()}"

# Use the ConcreteComponent and ConcreteDecorator classes to add behavior to the ConcreteComponent object
component = ConcreteComponent()
decorator_a = ConcreteDecoratorA(component)
decorator_b = ConcreteDecoratorB(decorator_a)

print(component.operation())
print(decorator_a.operation())
print(decorator_b.operation())
